File: frontend/src/utils/validators.ts
/**
 * 前端轻量校验（与后端 `app/core/validators.py` 保持同一套口径）。
 *
 * 为什么前后端各写一遍：前端这份只为「别让用户白等一次请求」。所以口径宁可按
 * **后端更宽**的那一档来，别出现「前端说格式不对、后端其实能收」这种自相矛盾的拦截。
 *
 * - 邮箱：正则与后端逐字一致——必须有且仅有一个 @，两侧非空且都不含空白，域名部分带点。
 * - 账号：3–30 位，仅 `a-z 0-9 _ -`，禁止 `@`。账号与邮箱共享同一个登录命名空间，
 *   禁止账号长得像邮箱能把「撞车」的面积从根上砍掉一大半（详见后端 validators.py）。
 */
const EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
const ACCOUNT_RE = /^[a-z0-9_-]{3,30}$/;

export const ACCOUNT_MIN_LEN = 3;
export const ACCOUNT_MAX_LEN = 30;

/** 账号名规则说明：与后端 `ACCOUNT_RULE_HINT` 同一句口径，避免两边提示不一致 */
export const ACCOUNT_RULE_HINT = "账号为 3–30 位，仅可包含字母、数字、下划线或中划线";

/** 去掉首尾空白并转小写（邮箱大小写不敏感，`A@x.com` 与 `a@x.com` 是同一个账号） */
export function normalizeEmail(raw: string): string {
  return raw.trim().toLowerCase();
}

export function isValidEmail(raw: string): boolean {
  return EMAIL_RE.test(normalizeEmail(raw));
}

/**
 * 账号名归一化：去空格 + 转小写。
 *
 * 与后端同口径——**只归一化不校验**，因为登录时对存量账号名必须宽容
 * （早期可能存过不合规的名字），格式规则只约束「新写入」的账号名。
 */
export function normalizeAccount(raw: string): string {
  return raw.trim().toLowerCase();
}

export function isValidAccount(raw: string): boolean {
  return ACCOUNT_RE.test(normalizeAccount(raw));
}

/**
 * 登录标识（账号或邮箱）的非空校验。
 *
 * 刻意**不校验格式**：两种标识的形态完全不同，用一个正则去卡只会误伤；
 * 填错了后端自然查不到，回一句「账号或密码错误」即可。
 */
export function isValidIdentifier(raw: string): boolean {
  return raw.trim().length > 0;
}

File: backend/app/core/validators.py
"""通用入参校验（集中一处，避免同一段校验在 auth / users / admin 里各写一遍）。

两套规则，职责不同：
- **邮箱**：够用就好。必须有且仅有一个 @，@ 两侧非空且都不含空白，域名部分带点。
  宁可放宽也不要误杀——一个邮箱是否真的可用，由「验证码能不能收到」来证明，
  正则再严也证明不了这一点。
- **账号名**：收紧，因为它是人自己起的登录标识。3–30 位、只允许 `a-z 0-9 _ -`，
  **明确禁止 `@`**。禁 @ 不是洁癖：账号名与邮箱共享同一个登录命名空间（登录时
  `WHERE username = ? OR email = ?`），禁止账号名长得像邮箱，能把「撞车」的面积
  从根上砍掉一大半。

两者都做「去空格 + 转小写」归一化：`A@X.com` 与 `a@x.com`、`ZhangSan` 与
`zhangsan` 都必须被当成同一个标识，否则用户换个大小写就注册出第二个号。
"""
import re

from app.core.exceptions import (
    ERR_ACCOUNT_INVALID,
    ERR_EMAIL_INVALID,
    BusinessError,
)

# ^本地部分@域名部分.顶级域  且全串无空白
_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")

# 账号名：3–30 位，小写字母 / 数字 / 下划线 / 中划线（归一化后只剩小写）
_ACCOUNT_RE = re.compile(r"^[a-z0-9_-]{3,30}$")

ACCOUNT_MIN_LEN = 3
ACCOUNT_MAX_LEN = 30

# 账号名的规则说明（前后端共用同一句口径，避免提示不一致）
ACCOUNT_RULE_HINT = "账号为 3–30 位，仅可包含字母、数字、下划线或中划线"


def normalize_email(raw: str | None) -> str:
    """统一去空格 + 转小写。"""
    return (raw or "").strip().lower()


def is_valid_email(raw: str | None) -> bool:
    return bool(_EMAIL_RE.match(normalize_email(raw)))


def require_valid_email(raw: str | None) -> str:
    """校验并返回归一化后的邮箱；不合法直接抛业务异常。"""
    email = normalize_email(raw)
    if not is_valid_email(email):
        raise BusinessError(ERR_EMAIL_INVALID, "邮箱格式不正确", 400)
    return email


def normalize_account(raw: str | None) -> str:
    """账号名归一化：去空格 + 转小写。

    只归一化、不校验：登录时对**存量账号**必须宽容（早期可能存过中文名、
    超短名），格式规则只约束「新写入」的账号名，否则一上线就把老账号锁在门外。
    """
    return (raw or "").strip().lower()


def is_valid_account(raw: str | None) -> bool:
    return bool(_ACCOUNT_RE.match(normalize_account(raw)))


def require_valid_account(raw: str | None) -> str:
    """校验并返回归一化后的账号名；不合法直接抛业务异常。"""
    account = normalize_account(raw)
    if not is_valid_account(account):
        raise BusinessError(ERR_ACCOUNT_INVALID, ACCOUNT_RULE_HINT, 400)
    return account


def normalize_identifier(raw: str | None) -> str:
    """登录 / 重置密码时的「账号或邮箱」标识。

    两种标识的归一化规则一致（去空格 + 转小写），所以这里不区分形态；
    登录时**刻意不做任何格式校验**——填错了自然查不到，回一句「账号或密码错误」
    即可，不必告诉对方「你这个格式不对」（那等于免费给出格式探测）。
    """
    return (raw or "").strip().lower()
